### PR TITLE
Implement setFetchReference - remove window dependency to support Node env

### PR DESCRIPTION
**Pull request by isBatak**
Fix for #17551

This PR removes window dependency inside `api-fetch` module to support fetch polyfill on Node, like `isomorphic-fetch`.

- [x] implement `setFetchReference` method for passing fetch polyfill
- [x] add README
- [ ] fix tests (need help with that)

Any ideas how to fix tests?

### CODE REVIEW

**Comment by gziolo** [comment]
@jsnajdr and @blowery - can you share your feedback given that you have much better knowledge about providing Node support for the existing code 🙇

**Comment by gziolo** [comment]
To fix unit tests you will need to use this new method rather than override `window.fetch` in the test:

https://github.com/WordPress/gutenberg/blob/72e35056c2dda55d46092380f16b349892dfa3b2/packages/api-fetch/src/test/index.js#L17-L21

In addition, all references to `window.fetch` should be replaced with the name of the variable which is used as mock.

**Comment by jsnajdr** [comment]
@sgomes is the main expert on polyfilling `fetch` 🙂

I don't think that the `setFetchReference` API is really necessary. The `isomorphic-fetch` package can be imported without using any of the exports:
```js
import 'isomorphic-fetch';
```
and for Node.js, it will add `fetch`, `Request`, `Response` and `Headers` functions to the `global` object.

Then the problem is not that `fetch` is not available as global, but that the global is different in different environments: `window` in browser document, `self` in workers, `global` in Node.

The standard solution for this is to add another standard global 🙂 called `globalThis`. We've been discussing exactly the same issue in Calypso with @sgomes: https://github.com/Automattic/wp-calypso/pull/37953#issuecomment-563217256

To summarize, the best solution would be:
- use the `isomorphic-fetch` package to get a polyfilled global `fetch` everywhere
- use the `globalThis` polyfill to get an universal `globalThis` reference everywhere
- use `globalThis.fetch` in the code -- it will work in all environments

**Comment by sgomes** [comment]
> To summarize, the best solution would be:
> 
> * use the `isomorphic-fetch` package to get a polyfilled global `fetch` everywhere
> * use the `globalThis` polyfill to get an universal `globalThis` reference everywhere
> * use `globalThis.fetch` in the code -- it will work in all environments

+1 on this solution. It's a pretty sane and explicit way of doing things.

There are one or two things to watch for, namely the fact that partial URLs are meaningless on the server; you want to ensure that any `fetch` that is intended to run on the server side uses absolute URLs. In the case of Calypso, this wasn't a problem in the application itself, but rather in its tests.

**Comment by Mamaduka** [comment]
The Node 18 shipped with [(experimental) fetch](https://nodejs.org/de/blog/announcements/v18-release-announce/#fetch-experimental). I hope this might make removing the `window` dependency for the package easier.

**Comment by youknowriad** [comment]
This seems like it's not relevant anymore, we can just use node.js fetch directly these days. Closing as this needs to be redone in a different way. Thanks for the efforts.
